framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,839 @@
+     * @param bottom the class of the bottom qualifier or {@code null}; if {@code null}, this method
+     *     has no effect
+     */
+    protected void specifyBottom(
+            Map<QualifierKind, Set<QualifierKind>> directSuperMap,
+            @Nullable Class<? extends Annotation> bottom) {
+        if (bottom != null) {
+            QualifierKind bottomKind = getQualifierKind(bottom.getCanonicalName());
+            if (bottomKind == null) {
+                throw new BugInCF(
+                        "QualifierKindHierarchy#specifyBottom: the given bottom class, %s, is not in the hierarchy.",
+                        bottom.getCanonicalName());
+            }
+            Set<QualifierKind> currentLeafs = new TreeSet<>(allQualifierKinds());
+            currentLeafs.remove(bottomKind);
+            directSuperMap.forEach(
+                    (sub, supers) -> {
+                        currentLeafs.removeAll(supers);
+                    });
+            Set<QualifierKind> bottomDirectSuperQuals = directSuperMap.get(bottomKind);
+            bottomDirectSuperQuals.addAll(currentLeafs);
+        }
+    }
+
+    /**
+     * Creates the set of top {@link QualifierKind}s by searching {@code directSuperMap} for
+     * qualifier kinds without any direct super qualifier kinds.
+     *
+     * @param directSuperMap a mapping from a {@link QualifierKind} to a set of its direct super
+     *     qualifier kinds; created by {@link #createDirectSuperMap()}
+     * @return the set of top {@link QualifierKind}s
+     */
+    // Subclasses should override createDirectSuperMap to change the tops and not this method,
+    // because other methods expect the directSuperMap to be complete.
+    private Set<QualifierKind> createTopsSet(
+            Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        Set<QualifierKind> tops = new TreeSet<>();
+        directSuperMap.forEach(
+                (qualifierKind, superQuals) -> {
+                    if (superQuals.isEmpty()) {
+                        tops.add(qualifierKind);
+                    }
+                });
+        return tops;
+    }
+
+    /**
+     * Creates the set of bottom {@link QualifierKind}s by searching {@code directSuperMap} for
+     * qualifiers that are not a direct super qualifier kind of another qualifier kind.
+     *
+     * @param directSuperMap a mapping from a {@link QualifierKind} to a set of its direct super
+     *     qualifier kinds; created by {@link #createDirectSuperMap()}
+     * @return the set of bottom {@link QualifierKind}s
+     */
+    // Subclasses should override createDirectSuperMap or specifyBottom to change the bottoms and
+    // not this method, because other methods expect the directSuperMap to be complete.
+    private Set<QualifierKind> createBottomsSet(
+            Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        Set<QualifierKind> bottoms = new TreeSet<>(directSuperMap.keySet());
```

Comment:
I wondered why this used a TreeSet.  It's not obviously more efficient.  Is it because you want the method to return the qualifiers in a deterministic order?  The Javadoc doesn't promise it.